[SPARK-29486][SQL] CalendarInterval should have 3 fields: months, days and microseconds

LinhongLiu opened the pull request:
### What changes were proposed in this pull request?
Current CalendarInterval has 2 fields: months and microseconds. This PR try to change it
to 3 fields: months, days and microseconds. This is because one logical day interval may
have different number of microseconds (daylight saving).


### Why are the changes needed?
One logical day interval may have different number of microseconds (daylight saving).
For example, in PST timezone, there will be 25 hours from 2019-11-2 12:00:00 to
2019-11-3 12:00:00


### Does this PR introduce any user-facing change?
no

### How was this patch tested?
unit test and new added test cases

LinhongLiu committed:
modify class CalendarInterval and CalendarIntervalSuite

LinhongLiu committed:
modify DateTimeUtils, DateTimeUtilsSuite, TemporalSequenceImpl

LinhongLiu committed:
modify CalendarInterval related classes in sql/catalyst

LinhongLiu committed:
fix test in sql/catalyst

LinhongLiu committed:
fix remaining tests

LinhongLiu committed:
fix failed tests

LinhongLiu committed:
code clean and more bug fix

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

LinhongLiu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala:
```diff
@@ -31,8 +31,9 @@ private object Triggers {
-      throw new IllegalArgumentException(s"Doesn't support month or year interval: $interval")
+    if (cal.months > 0 || cal.days > 0) {
+      throw new IllegalArgumentException(
+        s"Doesn't support day, week, month or year interval: $interval")
```
This will fail a current valid usage: `Trigger.ProcessingTime("10 day")`
Maybe we should still keep converting day interval and use a 24 hours = 1 day assumption

LinhongLiu commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala:
```diff
@@ -373,13 +373,39 @@ class DateTimeUtilsSuite extends SparkFunSuite with Matchers {
+    assert(timestampAddInterval(ts3, 36, 0, 123000, TimeZoneGMT.toZoneId) === ts5)
+  }
+
+  test("timestamp add days") {
```
test case to demonstrate the daylight saving case

LinhongLiu commented:
cc @xuanyuanking @cloud-fan

cloud-fan commented:
Let's create a ticket in https://issues.apache.org/jira/secure/Dashboard.jspa and put it in the title. You can follow other PRs: https://github.com/apache/spark/pulls

LinhongLiu commented:
> Let's create a ticket in https://issues.apache.org/jira/secure/Dashboard.jspa and put it in the title. You can follow other PRs: https://github.com/apache/spark/pulls

sorry, my mistake. ticket id is attached now.

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #112141 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112141/testReport)** for PR 26134 at commit [`3d954d6`](https://github.com/apache/spark/commit/3d954d6dbf72ae164e873b33f72d57c08361fead).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17135/
Test PASSed.

SparkQA commented:
**[Test build #112141 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112141/testReport)** for PR 26134 at commit [`3d954d6`](https://github.com/apache/spark/commit/3d954d6dbf72ae164e873b33f72d57c08361fead).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112141/
Test FAILed.

MaxGekk commented on common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java:
```diff
@@ -111,14 +111,13 @@ public static CalendarInterval fromCaseInsensitiveString(String s) {
     microseconds += toLong(m.group(8)) * MICROS_PER_MILLI;
     microseconds += toLong(m.group(9));
-    return new CalendarInterval((int) months, microseconds);
+    return new CalendarInterval((int) months, (int) days, microseconds);
```
I think it is better to catch int overflow by `Math.toIntExact` instead of silently ignore it.

LinhongLiu replied:
or maybe we can change this to long

MaxGekk replied:
Valid range of dates/timestamps is [0..10000). So, we should support days up to `10000 * 366 = 3660000`. Even `int` is so much but it is ok.

MaxGekk commented on common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java:
```diff
@@ -195,7 +194,7 @@ public static CalendarInterval fromDayTimeString(String s, String from, String t
       try {
         int sign = m.group(1) != null && m.group(1).equals("-") ? -1 : 1;
-        long days = m.group(2) == null ? 0 : toLongWithRange("day", m.group(3),
+        int days = m.group(2) == null ? 0 : (int) toLongWithRange("day", m.group(3),
```
Use `Math.toIntExact` and in other similar places

LinhongLiu replied:
can we use `Math.toIntExact` for only days field? as you know, there are long -> int in months filed as well, should we cover both of them?

cloud-fan replied:
This is OK here, we call `toLongWithRange` with `Integer.MAX_VALUE` so it's safe to cast to int.

MaxGekk commented on common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java:
```diff
@@ -332,31 +331,35 @@ public static long parseSecondNano(String secondNano) throws IllegalArgumentExce
+  public final int days;
   public final long microseconds;
 
   public long milliseconds() {
```
The method will return different values because you exclude days from microseconds

MaxGekk commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java:
```diff
@@ -132,16 +132,17 @@ private void writeUnalignedBytes(
-    grow(16);
+    grow(24);
 
     // Write the months and microseconds fields of Interval to the variable length portion.
```
and the days field

LinhongLiu replied:
thx

SparkQA commented:
**[Test build #112154 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112154/testReport)** for PR 26134 at commit [`3d954d6`](https://github.com/apache/spark/commit/3d954d6dbf72ae164e873b33f72d57c08361fead).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17146/
Test PASSed.

MaxGekk commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java:
```diff
@@ -132,16 +132,17 @@ private void writeUnalignedBytes(
     // Write the months and microseconds fields of Interval to the variable length portion.
     Platform.putLong(getBuffer(), cursor(), input.months);
-    Platform.putLong(getBuffer(), cursor() + 8, input.microseconds);
+    Platform.putLong(getBuffer(), cursor() + 8, input.days);
```
Just in case, why we put long if days is int?

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -2610,25 +2610,31 @@ object Sequence {
     private val backedSequenceImpl = new IntegralSequenceImpl[T](dt)
-    private val microsPerMonth = 28 * CalendarInterval.MICROS_PER_DAY
+    private val microsPerDay = 24 * CalendarInterval.MICROS_PER_HOUR
+    private val microsPerMonth = 28 * microsPerDay
```
Not related to the PR but it is interesting why 28 days per months here.

LinhongLiu replied:
the code will use microsPerMonth to estimate the array length first. use the minimum days in one month can make sure the array long enough. (smaller days means smaller steps in micro and means longer array length)

cloud-fan replied:
not related to this PR, but would be great if we can add a comment to explain it in the code.

MaxGekk commented on sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out:
```diff
@@ -238,7 +238,7 @@ from interval_arithmetic
 struct<tsval:timestamp,CAST(tsval - interval 14 weeks 1 days 11 hours 22 minutes 33 seconds 123 milliseconds 456 microseconds AS TIMESTAMP):timestamp,CAST(tsval - interval -14 weeks -1 days -11 hours -22 minutes -33 seconds -123 milliseconds -456 microseconds AS TIMESTAMP):timestamp,CAST(tsval + interval 14 weeks 1 days 11 hours 22 minutes 33 seconds 123 milliseconds 456 microseconds AS TIMESTAMP ...
 -- !query 17 output
-2012-01-01 00:00:00	2011-09-23 13:37:26.876544	2012-04-09 12:22:33.123456	2012-04-09 12:22:33.123456	2011-09-23 13:37:26.876544	2011-09-23 13:37:26.876544	2012-04-09 12:22:33.123456
+2012-01-01 00:00:00	2011-09-23 12:37:26.876544	2012-04-09 11:22:33.123456	2012-04-09 11:22:33.123456	2011-09-23 12:37:26.876544	2011-09-23 12:37:26.876544	2012-04-09 11:22:33.123456
```
Why it there difference of 1 hour?

LinhongLiu replied:
as you know it's caused by daylight saving

MaxGekk commented on common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java:
```diff
@@ -365,12 +368,12 @@ public boolean equals(Object other) {
   @Override
   public int hashCode() {
-    return 31 * months + (int) microseconds;
+    return 31 * (31 * months + days) + (int) microseconds;
```
Not related to the PR but I am wondering why the most significant bits of microseconds are completely excluded from the hash.

MaxGekk commented:
> Why are the changes needed?
> One logical day interval may have different number of microseconds (daylight saving).
For example, in PST timezone, there will be 25 hours from 2019-11-2 12:00:00 to
2019-11-3 12:00:00

Yes, actual duration of the daylight saving day can be 25 hours but logically any day has 24 hours. Spark `INTERVAL` type defines not duration, it is adjustment of local date-time. After your changes, it adjusts logical (local) months, days and for some reasons physical time duration of a day. It looks like inconsistent, doesn't it?

As you point out, the daylight saving in PST will happen at `2019-11-03 02:00:00`. So, the next timestamp with granularity of 1 microsecond is `2019-11-03 01:00:01`. Even more, the local timestamp `2019-11-03 02:00:00` will be showed on local clock twice.
At the first time:
```
timestamp'2019-11-03 02:00:00' + interval 1 microsecond = timestamp'2019-11-03 01:00:00.000001'
```
At the second time:
```
timestamp'2019-11-03 02:00:00' + interval 1 microsecond = timestamp'2019-11-03 02:00:00.000001'
```
Could you elaborate how your changes resolve this situation?

cloud-fan commented:
one day is always 24 hours, but one day interval may not. Daylight saving doesn't change the # of hours in a day, but the timezone offset of a region. For example, California is in UTC-7 right now, but will be in UTC-8 after 3 Nov 2019. So adding one day to `3 Nov 2019` at local time maybe different from adding 24 hours.

However, I'd like to know if it's allowed to set timezone to a region name that has undetermined timezone offset. If it's not allowed then one day interval is always 24 hours in SQL.

MaxGekk commented:
> ... but one day interval may not.

As I can understand, new `microseconds` field means actual time duration inside of a day, correct? Back to the example, it means:
```
timestamp'2019-11-03 01:59:59 PST' + interval 2 seconds = timestamp'2019-11-03 01:00:01 PST'
```
and to get `timestamp'2019-11-03 02:00:01 PST'`, we should add another interval:
```
timestamp'2019-11-03 01:59:59 PST' + interval 1 hour 2 seconds = timestamp'2019-11-03 02:00:01 PST'
```

> However, I'd like to know if it's allowed to set timezone to a region name that has undetermined timezone offset.

For example, Java 8 time API allows to get 0, 1 or 2 zone offsets depend on local timestamp, see https://docs.oracle.com/javase/8/docs/api/java/time/zone/ZoneRules.html#getValidOffsets-java.time.LocalDateTime- , see:
```
The mapping from a local date-time to an offset is not straightforward. There are three cases:
 -   Normal, with one valid offset. For the vast majority of the year, the normal case applies, where there is a single valid offset for the local date-time.
 -   Gap, with zero valid offsets. This is when clocks jump forward typically due to the spring daylight savings change from "winter" to "summer". In a gap there are local date-time values with no valid offset.
 -   Overlap, with two valid offsets. This is when clocks are set back typically due to the autumn daylight savings change from "summer" to "winter". In an overlap there are local date-time values with two valid offsets.
```


> If it's not allowed then one day interval is always 24 hours in SQL.

The SQL standard defines time zones as TIMEZONE_HOUR:TIMEZONE_MINUTE. Due to this, there is no issue of converting zone id (region name) to offsets.

SparkQA commented:
**[Test build #112154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112154/testReport)** for PR 26134 at commit [`3d954d6`](https://github.com/apache/spark/commit/3d954d6dbf72ae164e873b33f72d57c08361fead).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112154/
Test FAILed.

MaxGekk commented:
Since SQL has always determined zone offsets, a day has 24 hours always, and interval hours are in `0..23`:
<img width="422" alt="Screen Shot 2019-10-16 at 13 48 43" src="https://user-images.githubusercontent.com/1580697/66912682-b4d2ae00-f01b-11e9-98d1-fad089969fbf.png">

Current implementation of `CalendarIntervalType` is correct union of SQL year-month INTERVAL and day-time INTERVAL. Also there is no issue of adding intervals to timestamps on Daylight saving day if we assume that the time part of intervals is just "logical" adjustment.

Taking into account that local hours of timestamps are always in the range of `[0, 23]`, @LinhongLiu could you show concrete example when old (current) implementation returns wrong results, and your changes fix that. And it would be nice to put such example in the description of this PR.

LinhongLiu commented:
@MaxGekk @cloud-fan thanks for your review, here is my answer for the questions:

> local timestamp 2019-11-03 02:00:00 will be showed on local clock twice

my solution won't fix this, but the original code won't, too. actually I don't think this matters.


> could you show concrete example when old (current) implementation returns wrong results, and your changes fix that

the first case is adding `1 day` on `2019-11-02 12:00:00 PST` in current code will result in `2019-11-03 11:00:00 PST`. But as you said, this won't produce a wrong results.
the second case is in SS, if we do a window agg like this:
```
df.groupBy(window($"time", "1 day")).sum
```
Since this function is considered timezone, in current implementation, we may have window like this:
`2019-11-01 00:00:00` - `2019-11-01 23:59:59`
`2019-11-02 00:00:00` - `2019-11-02 23:59:59`
`2019-11-03 00:00:00` - `2019-11-03 22:59:59`
`2019-11-03 23:00:00` - `2019-11-04 22:59:59`
I think the user don't wan't see aggregation in this kind of window.

xuanyuanking commented on common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java:
```diff
@@ -365,12 +368,12 @@ public boolean equals(Object other) {
 
     CalendarInterval o = (CalendarInterval) other;
-    return this.months == o.months && this.microseconds == o.microseconds;
+    return this.months == o.months && this.days == o.days && this.microseconds == o.microseconds;
```
Is this line changed manually? We'd better use IDE to generate the equals and hashCode.

LinhongLiu committed:
address comments and fix tests

LinhongLiu committed:
Merge remote-tracking branch 'origin/master' into calendarinterval

MaxGekk commented:
> the first case is adding 1 day on 2019-11-02 12:00:00 PST in current code will result in 2019-11-03 11:00:00 PST

ok. Your PR changes this behavior. Not sure what users want to see - the timestamp after exactly 24 hours or 25 hours. Also it is debatable that it is worth to consume 30% more memory for this.

> the second case is in SS, if we do a window agg like this:
> df.groupBy(window($"time", "1 day")).sum

This is actually behavior change, and users will struggle trouble shooting such difference in results, I think. When they usually summarize over 24 hours, but sometimes over 23 or 25 hours after your changes. Not sure that it will be expected behavior.

MaxGekk commented:
> @MaxGekk do you think Spark violates the SQL standard if ZoneId is used to represent the session local timezone?

@cloud-fan I think so. If we accept the session time zone as zone offset (`TIMEZONE_HOUR:TIMEZONE_MINUTE`) only some problems will gone for sure. And this will speed up date-time operations significantly because conversion of `ZoneId` -> `ZoneOffset` is very time consuming. It has to look up at historical rules, for instance: http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/time/zone/ZoneRules.java#l633

> I see many datetime operations in DateTimeUtils are using ZoneId.

ZoneId itself is not the problem because ZoneOffset is sub-type of ZoneId and Java 8 time functions can optimally handle ZoneId instances when they are ZoneOffset actually. Theoretically, we could require `spark.sql.session.timeZone` to be set as zone offset `TIMEZONE_HOUR:TIMEZONE_MINUTE` but this will break user apps highly likely.

SparkQA commented:
**[Test build #112173 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112173/testReport)** for PR 26134 at commit [`4c31401`](https://github.com/apache/spark/commit/4c31401f05c22041a3727cf39b4bcef918e42124).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  val thresholds: DoubleArrayParam = new DoubleArrayParam(this, \"thresholds\", \"Thresholds in multi-class classification to adjust the probability of predicting each class. Array must have length equal to the number of classes, with values > 0 excepting that at most one value may be 0. The class with largest value p/t is predicted, where p is the original probability of that class and t is the class's threshold\", (t: Array[Double]) => t.forall(_ >= 0) && t.count(_ == 0) <= 1)`
  * `class Binarizer(JavaTransformer, HasThreshold, HasThresholds, HasInputCol, HasOutputCol,`
  * `case class SubtractTimestamps(endTimestamp: Expression, startTimestamp: Expression)`
  * `case class SubtractDates(left: Expression, right: Expression)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112173/
Test FAILed.

MaxGekk commented:
Just in case, PostgreSQL represents intervals as this PR proposes: https://github.com/postgres/postgres/blob/bffe1bd68457e43925c362d8728ce3b25bdf1c94/src/include/datatype/timestamp.h#L43-L49

LinhongLiu commented:
@MaxGekk 
I think both of us agreed now that undetermined zone offset can cause ambiguity if user want add a 'interval 1 day' and different user may want different time shift ( 24 hours or 25 hours)

My PR change current behavior but it gives the user a choice to use a physical 1 day (23,24,25 hours maybe) or use a logical 1 day (24 hours exactly, this is consistent to current behavior). This change can reduce the ambiguity.

At last, about the 30% memory consume, in my opinion, CalendarInterval usually is a constant value and used to shift a column of timestamp type. This means we rarely save CalendarInterval as a column of a table. So memory consuming is not really a big problem.

cloud-fan commented:
I think the session local timezone can still be a `ZoneId`, but we should only resolve `ZoneId` to `ZoneOffset` once for a query. It's super confusing that the `ZoneOffset` changes in the middle of a query. Many problems can go away if we fix it.

MaxGekk commented:
> we should only resolve ZoneId to ZoneOffset once for a query.

@cloud-fan How? To convert `ZoneId` to `ZoneOffset`, you have to point out an `Instant` or `LacalDateTime`. Which one out of bunch of values in your dataset are you going to use in the conversion?

cloud-fan commented:
To be more clear, the timestamp in Spark is `timestamp + zoneId`, and we need extra effort to resolve the zoneId to zoneOffset according to the associated timestamp value.

This doesn't follow the SQL standard. We should define the timestamp type as `timestamp + zoneOffset`. That said, the session local timezone should be a zoneOffset.

Users should still be able to specify any zoneId as session local timezone, for backward compatibility. We should convert zoneId to zoneOffset at the beginning of a query, according to the current time of query submission.

MaxGekk commented:
> We should convert zoneId to zoneOffset at the beginning of a query, according to the current time of query submission.

I cannot agree with you. For example, time zone rules were changed twice in Russia during last 10 years (https://en.wikipedia.org/wiki/Time_in_Russia#Daylight_saving_time). If I have timestamps < 2011, between 2011 and 2014, > 2014 year, processing timestamp before 2014 years give me wrong results. Actually, you are going to force me to split my query to many separate queries per each daylight saving rule otherwise I will get incorrect results where local timestamps are involved.

Even if daylight saving rules have been not changed in some region, how should users process their datasets with timestamps belong to winter and summer periods? Let's say an user applies the `hour()` function some winter timestamps but current date belongs to summer period.

> This doesn't follow the SQL standard. We should define the timestamp type as timestamp + zoneOffset

To be more clear:
In Spark, timestamp is instant + session zone id
In SQL, timestamp with time zone is local date-time + zone offset per each value.

The key thing is per each value. The SQL standard just moved out the scope resolving ZoneId -> ZoneOffsets.

cloud-fan commented:
> In Spark, timestamp is instant + session zone id

I don't fully agree. Timestamp in Spark is instant. When Spark parses string to timestamp(e.g. `TIMESTAMP '2019-1-1 12:12:12'`), it assumes the input string is local datetime(unless it carries timezone) and parses it using the session local timezone id. That's why Spark stores timestamp as microseconds from UTC epoch: because it's an instant.

The session local timezone id is also used when Spark needs to know the datetime fields of the instant, e.g. to string, extracting field functions, adding interval, etc. That said, a lot of datetime operations in Spark rely on the session local timezone id.

Now we have 2 choices:
1. keep session local timezone id as it is. This raises the special requirement to have a day field in interval.
2. change session local timezone id to zone offset. This makes some datetime functions hard to use for some use cases.

TIMESTAMP WITH LOCAL TIMEZONE is not a standard SQL type, so we are on our own now. To be honest I'd like to make Spark follow SQL standard and use TIMESTAMP WITHOUT TIMEZONE, but it changes the semantic and we don't know how to deal with existing timestamp data.

I'm really on the fence. @MaxGekk do you have some better ideas?

MaxGekk commented:
> I'm really on the fence. @MaxGekk do you have some better ideas?

I would strictly follow the SQL standard and implement SQL `TIMESTAMP WITHOUT TZ` and `TIMESTAMP WITH TZ` (with TZ_HOUR:TZ_MINUTE). This could be done when `spark.sql.ansi.enabled = true`, or independently from the setting.

For our default settings at the moment `spark.sql.ansi.enabled = false`, I would **not** change the session local time zone because we definitely break existing user apps. Changing structure of `INTERVAL` type looks reasonable. 

> To be honest I'd like to make Spark follow SQL standard and use TIMESTAMP WITHOUT TIMEZONE, but it changes the semantic and we don't know how to deal with existing timestamp data.

Spark's `TIMESTAMP` and SQL `TIMESTAMP WITHOUT TIMEZONE` is different types, no doubt. I don't think we can smoothly replace one by another one. I imagine the types as similar Java 8 classes:
Spark's `TIMESTAMP` - `java.time.Instant`
SQL `TIMESTAMP WITHOUT TZ` - `java.time.LocalDateTime`
SQL `TIMESTAMP WITH TZ` - `java.time.ZonedDateTime`

As Java 8 time API has all 3 types, like that Spark could have all 3.

cloud-fan commented:
> Changing structure of INTERVAL type looks reasonable.

So let's move forward with this PR.

There is already a proposal in the dev list to add TIMESTAMP WITHOUT TZ and TIMESTAMP WITH TZ, let's wait for it.

LinhongLiu committed:
fix python

LinhongLiu committed:
Merge remote-tracking branch 'origin/master' into calendarinterval

SparkQA commented:
**[Test build #112281 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112281/testReport)** for PR 26134 at commit [`211543f`](https://github.com/apache/spark/commit/211543f2bb3d2a219501624f9f44f48df2c9a6bd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17264/
Test PASSed.

LinhongLiu commented:
Hi @cloud-fan and @MaxGekk ，thanks and sorry for the late reply. I read through the discussions and learnt a lot. I updated the patch and fixed failed test cases. Since adding a day field to INTERVAL type seems to be a reasonable way to solve the ambiguity caused session local time, shell we move forward and give some advice to this PR?

LinhongLiu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:
```diff
@@ -103,9 +103,9 @@ object TimeWindow {
+    if (cal.months > 0 || cal.days > 0) {
       throw new IllegalArgumentException(
-        s"Intervals greater than a month is not supported ($interval).")
+        s"Intervals greater than a day is not supported ($interval).")
```
`TimeWindow` use `Long` to represent `window, slide, startTime`, if we change `Long` to `CalendarInterval`, we can reduce this constraint

SparkQA commented:
**[Test build #112281 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112281/testReport)** for PR 26134 at commit [`211543f`](https://github.com/apache/spark/commit/211543f2bb3d2a219501624f9f44f48df2c9a6bd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `            s\"Failed to get main class in JAR with error '$`
  * `        error(\"No main class set in JAR; please specify one with --class.\")`
  * `abstract class EventLogFileReader(`
  * `class SingleFileEventLogFileReader(`
  * `class RollingEventLogFilesFileReader(`
  * `abstract class EventLogFileWriter(`
  * `class SingleEventLogFileWriter(`
  * `class RollingEventLogFilesWriter(`
  * `  class DriverEndpoint extends IsolatedRpcEndpoint with Logging `
  * `case class ShuffleBlockBatchId(`
  * `        require(probability != null, \"probability of each class can not be null\")`
  * `class _LinearSVCParams(_JavaClassifierParams, HasRegParam, HasMaxIter, HasFitIntercept, HasTol,`
  * `class LinearSVC(JavaClassifier, _LinearSVCParams, JavaMLWritable, JavaMLReadable):`
  * `class LinearSVCModel(JavaClassificationModel, _LinearSVCParams, JavaMLWritable, JavaMLReadable):`
  * `class _LogisticRegressionParams(_JavaProbabilisticClassifierParams, HasRegParam,`
  * `class LogisticRegression(JavaProbabilisticClassifier, _LogisticRegressionParams, JavaMLWritable,`
  * `class LogisticRegressionModel(JavaProbabilisticClassificationModel, _LogisticRegressionParams,`
  * `class _NaiveBayesParams(_JavaPredictorParams, HasWeightCol):`
  * `class NaiveBayes(JavaProbabilisticClassifier, _NaiveBayesParams, HasThresholds, HasWeightCol,`
  * `class NaiveBayesModel(JavaProbabilisticClassificationModel, _NaiveBayesParams, JavaMLWritable,`
  * `class _MultilayerPerceptronParams(_JavaProbabilisticClassifierParams, HasSeed, HasMaxIter,`
  * `class MultilayerPerceptronClassifier(JavaProbabilisticClassifier, _MultilayerPerceptronParams,`
  * `class _LinearRegressionParams(_JavaPredictorParams, HasRegParam, HasElasticNetParam, HasMaxIter,`
  * `class LinearRegression(JavaPredictor, _LinearRegressionParams, JavaMLWritable, JavaMLReadable):`
  * `class LinearRegressionModel(JavaPredictionModel, _LinearRegressionParams, GeneralJavaMLWritable,`
  * `class _IsotonicRegressionParams(HasFeaturesCol, HasLabelCol, HasPredictionCol, HasWeightCol):`
  * `class IsotonicRegression(JavaEstimator, _IsotonicRegressionParams, HasWeightCol,`
  * `class IsotonicRegressionModel(JavaModel, _IsotonicRegressionParams, JavaMLWritable,`
  * `class _GeneralizedLinearRegressionParams(_JavaPredictorParams, HasFitIntercept, HasMaxIter,`
  * `class GeneralizedLinearRegression(JavaPredictor, _GeneralizedLinearRegressionParams,`
  * `class GeneralizedLinearRegressionModel(JavaPredictionModel, _GeneralizedLinearRegressionParams,`
  * `case class BitwiseCount(child: Expression) extends UnaryExpression with ExpectsInputTypes `
  * `case class AnalyzeTableStatement(`
  * `case class AnalyzeColumnStatement(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112281/
Test PASSed.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala:
```diff
@@ -256,12 +256,12 @@ object StreamingJoinHelper extends PredicateHelper with Logging {
             case CalendarIntervalType =>
               val calendarInterval = lit.value.asInstanceOf[CalendarInterval]
-              if (calendarInterval.months > 0) {
+              if (calendarInterval.months > 0 || calendarInterval.days > 0) {
```
This can be breaking change. Why did you reject interval with days? By the way, `days` can be converted to microseconds

LinhongLiu replied:
done, I restore the constraint back

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:
```diff
@@ -103,9 +103,9 @@ object TimeWindow {
   private def getIntervalInMicroSeconds(interval: String): Long = {
     val cal = CalendarInterval.fromCaseInsensitiveString(interval)
-    if (cal.months > 0) {
+    if (cal.months > 0 || cal.days > 0) {
```
Intervals with days could be converted to microseconds.

MaxGekk commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -731,7 +731,7 @@ class Dataset[T] private[sql](
             cause = Some(e))
       }
-    require(parsedDelay.milliseconds >= 0 && parsedDelay.months >= 0,
+    require(parsedDelay.milliseconds >= 0 && parsedDelay.days >= 0 && parsedDelay.months >= 0,
```
Not related to the PR but `interval 1 month -1 day -1 millisecond` doesn't mean that the interval is negative.

MaxGekk replied:
The PR https://github.com/apache/spark/pull/26177 to fix this

MaxGekk commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/GroupStateImpl.scala:
```diff
@@ -160,12 +160,12 @@ private[sql] class GroupStateImpl[S] private(
   private def parseDuration(duration: String): Long = {
     val cal = CalendarInterval.fromCaseInsensitiveString(duration)
-    if (cal.milliseconds < 0 || cal.months < 0) {
+    if (cal.milliseconds < 0 || cal.days < 0 || cal.months < 0) {
```
Again even if `day` is negative, the duration can be positive.

MaxGekk commented:
Please, rebase on the recent master. Your changes can affect https://github.com/apache/spark/commit/77fe8a8e7cdb0eae9f66b2d10c0d35f784663547#diff-eba257f41b49f470321579875f054f00R85-R90

LinhongLiu committed:
use 24 hours = 1 day assumption in window, trigger and watermark

LinhongLiu committed:
Merge remote-tracking branch 'origin/master' into calendarinterval

SparkQA commented:
**[Test build #112329 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112329/testReport)** for PR 26134 at commit [`3a6518a`](https://github.com/apache/spark/commit/3a6518ac5828d826f65b0e1a66be8c85c2935321).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17308/
Test PASSed.

LinhongLiu committed:
streaming changes followup

SparkQA commented:
**[Test build #112330 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112330/testReport)** for PR 26134 at commit [`4d8383c`](https://github.com/apache/spark/commit/4d8383c3430db4a0ef2bff3eb93bf80b0c8cdd1c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17309/
Test PASSed.

SparkQA commented:
**[Test build #112329 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112329/testReport)** for PR 26134 at commit [`3a6518a`](https://github.com/apache/spark/commit/3a6518ac5828d826f65b0e1a66be8c85c2935321).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class ExtractIntervalPart(`
  * `case class ExtractIntervalMillenniums(child: Expression)`
  * `case class ExtractIntervalCenturies(child: Expression)`
  * `case class ExtractIntervalDecades(child: Expression)`
  * `case class ExtractIntervalYears(child: Expression)`
  * `case class ExtractIntervalQuarters(child: Expression)`
  * `case class ExtractIntervalMonths(child: Expression)`
  * `case class ExtractIntervalDays(child: Expression)`
  * `case class ExtractIntervalHours(child: Expression)`
  * `case class ExtractIntervalMinutes(child: Expression)`
  * `case class ExtractIntervalSeconds(child: Expression)`
  * `case class ExtractIntervalMilliseconds(child: Expression)`
  * `case class ExtractIntervalMicroseconds(child: Expression)`
  * `case class ExtractIntervalEpoch(child: Expression)`
  * `case class RepairTableStatement(tableName: Seq[String]) extends ParsedStatement`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112329/
Test FAILed.

SparkQA commented:
**[Test build #112330 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112330/testReport)** for PR 26134 at commit [`4d8383c`](https://github.com/apache/spark/commit/4d8383c3430db4a0ef2bff3eb93bf80b0c8cdd1c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112330/
Test FAILed.

LinhongLiu committed:
fix conflict

SparkQA commented:
**[Test build #112342 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112342/testReport)** for PR 26134 at commit [`1ac157e`](https://github.com/apache/spark/commit/1ac157eab2467b1b467d92ad35a60ddc0cf8dc11).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17321/
Test PASSed.

SparkQA commented:
**[Test build #112342 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112342/testReport)** for PR 26134 at commit [`1ac157e`](https://github.com/apache/spark/commit/1ac157eab2467b1b467d92ad35a60ddc0cf8dc11).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112342/
Test PASSed.

MaxGekk commented:
> There is already a proposal in the dev list to add TIMESTAMP WITHOUT TZ and TIMESTAMP WITH TZ, let's wait for it.

@cloud-fan Is it in public dev list http://apache-spark-developers-list.1001551.n3.nabble.com/ ? I haven't found this proposal so far.

LinhongLiu commented:
ping @cloud-fan. what's your suggestions about this PR?

cloud-fan commented:
I'll start reviewing it after the conflicts are resolved. Sorry for the delay!

cloud-fan commented:
@MaxGekk I'm not sure why we can't find it on the dev-list website, but I can find it when searching dev list emails. This is the design doc: https://docs.google.com/document/d/1gNRww9mZJcHvUDCXklzjFEQGpefsuR_akCDfWsdE35Q/edit?usp=sharing

LinhongLiu committed:
Merge remote-tracking branch 'origin/master' into calendarinterval

SparkQA commented:
**[Test build #112708 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112708/testReport)** for PR 26134 at commit [`aac92bd`](https://github.com/apache/spark/commit/aac92bd2f4216a2124e69396d96bbcdbfe72ae3d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17656/
Test PASSed.

SparkQA commented:
**[Test build #112708 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112708/testReport)** for PR 26134 at commit [`aac92bd`](https://github.com/apache/spark/commit/aac92bd2f4216a2124e69396d96bbcdbfe72ae3d).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112708/
Test FAILed.

cloud-fan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17704/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java:
```diff
@@ -132,16 +132,17 @@ private void writeUnalignedBytes(
   public final void write(int ordinal, CalendarInterval input) {
     // grow the global buffer before writing data.
-    grow(16);
+    grow(24);
```
it's int, int, long, so 16 bytes should be enough.

SparkQA commented:
**[Test build #112760 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112760/testReport)** for PR 26134 at commit [`aac92bd`](https://github.com/apache/spark/commit/aac92bd2f4216a2124e69396d96bbcdbfe72ae3d).

cloud-fan commented on sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java:
```diff
@@ -132,16 +132,17 @@ private void writeUnalignedBytes(
 
-    // Write the months and microseconds fields of Interval to the variable length portion.
+    // Write the months, days and microseconds fields of Interval to the variable length portion.
     Platform.putLong(getBuffer(), cursor(), input.months);
```
we write `months` as long to align with word(8 bytes) boundary. Now we have an extra int in the interval, so padding is not needed now.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:
```diff
@@ -573,13 +573,15 @@ object DateTimeUtils {
-      months: Int,
-      microseconds: Long,
-      zoneId: ZoneId): SQLTimestamp = {
+    start: SQLTimestamp,
```
nit: wrong indentation

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:
```diff
@@ -61,7 +61,7 @@ object IntervalUtils {
   }
 
   def getDays(interval: CalendarInterval): Long = {
```
We need to redefine the semantic here. To exact days from `INTERVAL 1 day 48 hour`, previously we return 3 days, now we return 1 day. Which one is corrected? cc @MaxGekk

cloud-fan replied:
Same question to `getHours`. For now it's not consistent with `getDays`

MaxGekk replied:
The method was added for feature parity with PostgreSQL which tracks days in a separate variable too, and doesn't perform normalization:
```sql
maxim=# select INTERVAL '1 day 48 hour';
    interval
----------------
 1 day 48:00:00
(1 row)
```
> Which one is corrected?

I think we could change current behavior.

cloud-fan replied:
I played with pgsql a little bit
```
cloud0fan=# SELECT EXTRACT(DAY FROM INTERVAL '1 days 50 hours');
 date_part 
-----------
         1
(1 row)

cloud0fan=# SELECT EXTRACT(HOUR FROM INTERVAL '1 days 50 hours');
 date_part 
-----------
        50
(1 row)

cloud0fan=# SELECT EXTRACT(HOUR FROM INTERVAL '1 days 50 hours 100 minutes');
 date_part 
-----------
        51
(1 row)
```

So `getDays` should return the `days` field directly. `getHours` should simply return `microseconds / MICROS_PER_HOUR`. @LinhongLiu can you fix it?

LinhongLiu committed:
fix code sytle and test cases

LinhongLiu committed:
Merge remote-tracking branch 'origin/master' into calendarinterval

LinhongLiu committed:
change long,long,long to int,int,long when store CanlendarInterval in UnsafeRow

LinhongLiu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala:
```diff
@@ -107,7 +108,7 @@ object TimeWindow {
         s"Intervals greater than a month is not supported ($interval).")
     }
-    cal.microseconds
+    cal.days * CalendarInterval.MICROS_PER_DAY + cal.microseconds
```
@cloud-fan here we still use a "1 day = 24 hours" assumption. is this OK?

cloud-fan replied:
here we use interval as a duration, and it's safe to use `1 day = 24 hours` (this is always true).

SparkQA commented:
**[Test build #112760 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112760/testReport)** for PR 26134 at commit [`aac92bd`](https://github.com/apache/spark/commit/aac92bd2f4216a2124e69396d96bbcdbfe72ae3d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112760/
Test FAILed.

SparkQA commented:
**[Test build #112765 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112765/testReport)** for PR 26134 at commit [`2edd8a0`](https://github.com/apache/spark/commit/2edd8a0cffaf004b1c9c3f5ecfa5346736107b82).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17710/
Test PASSed.

SparkQA commented:
**[Test build #112765 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112765/testReport)** for PR 26134 at commit [`2edd8a0`](https://github.com/apache/spark/commit/2edd8a0cffaf004b1c9c3f5ecfa5346736107b82).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112765/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -2610,25 +2610,33 @@ object Sequence {
-        // about a month length in microseconds
-        val intervalStepInMicros = stepMicros + stepMonths * microsPerMonth
+        // about a month length in days and a day length in microseconds
+        // We choose a minimum days(28) in one month to make sure the estimated
```
can we move this comment to https://github.com/apache/spark/pull/26134/files#diff-9853dcf5ce3d2ac1e94d473197ff5768R2617 ?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:
```diff
@@ -87,6 +87,7 @@ object IntervalUtils {
     var result = interval.microseconds
+    result += DateTimeUtils.MICROS_PER_DAY * interval.days
     result += MICROS_PER_YEAR * (interval.months / MONTHS_PER_YEAR)
     result += MICROS_PER_MONTH * (interval.months % MONTHS_PER_YEAR)
```
not related to this PR, but isn't it simply `result += MICROS_PER_MONTH * interval.months`?

MaxGekk replied:
It could be if MICROS_PER_YEAR = 12 * MICROS_PER_MONTH . I strictly followed PostgreSQL implementation: https://github.com/postgres/postgres/blob/bffe1bd68457e43925c362d8728ce3b25bdf1c94/src/backend/utils/adt/timestamp.c#L5016-L5022

LinhongLiu committed:
fix comment

SparkQA commented:
**[Test build #112813 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112813/testReport)** for PR 26134 at commit [`336f0db`](https://github.com/apache/spark/commit/336f0db576b99f9524e5893064ccd393d5b0432a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17751/
Test PASSed.

SparkQA commented:
**[Test build #112814 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112814/testReport)** for PR 26134 at commit [`4e97bc0`](https://github.com/apache/spark/commit/4e97bc05f35fc9ddb05e945d1fe1a83f66abd0c7).

SparkQA commented:
**[Test build #112813 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112813/testReport)** for PR 26134 at commit [`336f0db`](https://github.com/apache/spark/commit/336f0db576b99f9524e5893064ccd393d5b0432a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112813/
Test PASSed.

SparkQA commented:
**[Test build #112814 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112814/testReport)** for PR 26134 at commit [`4e97bc0`](https://github.com/apache/spark/commit/4e97bc05f35fc9ddb05e945d1fe1a83f66abd0c7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112814/
Test PASSed.

LinhongLiu committed:
Merge remote-tracking branch 'origin/master' into calendarinterval

SparkQA commented:
**[Test build #113013 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113013/testReport)** for PR 26134 at commit [`0e87e2d`](https://github.com/apache/spark/commit/0e87e2dc85fd2fcac3789fa6ada9d5ed78cde253).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17935/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:
```diff
@@ -1205,14 +1205,14 @@ case class TimeSub(start: Expression, interval: Expression, timeZoneId: Option[S
     val itvl = interval.asInstanceOf[CalendarInterval]
     DateTimeUtils.timestampAddInterval(
-      start.asInstanceOf[Long], 0 - itvl.months, 0 - itvl.microseconds, zoneId)
+      start.asInstanceOf[Long], 0 - itvl.months, 0 - itvl.days, 0 - itvl.microseconds, zoneId)
```
can we change `timestampAddInterval` to accept `CalendarInterval` as parameter? Then the caller side can be simplified: `DateTimeUtils.timestampAddInterval(start, itvl.negative())` or `DateTimeUtils.timestampAddInterval(start, itvl)`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala:
```diff
@@ -63,10 +63,10 @@ case class ExtractIntervalMonths(child: Expression)
 
 case class ExtractIntervalHours(child: Expression)
-    extends ExtractIntervalPart(child, ByteType, getHours, "getHours")
+    extends ExtractIntervalPart(child, LongType, getHours, "getHours")
```
Why the hours need to be long type now?

SparkQA commented:
**[Test build #113013 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113013/testReport)** for PR 26134 at commit [`0e87e2d`](https://github.com/apache/spark/commit/0e87e2dc85fd2fcac3789fa6ada9d5ed78cde253).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait HasRelativeError extends Params `
  * `class _ImputerParams(HasInputCol, HasInputCols, HasOutputCol, HasOutputCols, HasRelativeError):`
  * `class _OneHotEncoderParams(HasInputCol, HasInputCols, HasOutputCol, HasOutputCols,`
  * `class _RobustScalerParams(HasInputCol, HasOutputCol, HasRelativeError):`
  * `class HasRelativeError(Params):`
  * `  class ResolveCoalesceHints(conf: SQLConf) extends Rule[LogicalPlan] `
  * `case class AlterTableRecoverPartitionsStatement(`
  * `case class DropNamespaceStatement(`
  * `case class LoadDataStatement(`
  * `case class ShowColumnsStatement(`
  * `case class DropNamespace(`
  * `case class DropNamespaceExec(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113013/
Test FAILed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:
```diff
@@ -234,11 +235,10 @@ object IntervalUtils {
       micros = Math.addExact(micros, Math.multiplyExact(minutes, MICROS_PER_MINUTE))
       micros = Math.addExact(micros, Math.multiplyExact(seconds, DateTimeUtils.MICROS_PER_SECOND))
-      new CalendarInterval(0, sign * micros)
+      new CalendarInterval(0, sign * days.toInt, sign * micros)
```
nit: we can cast to int at the beginning
```
    val days = if (m.group(2) == null) {
        0
      } else {
        toLongWithRange("day", m.group(3), 0, Integer.MAX_VALUE).toInt
      }
```

LinhongLiu replied:
sure, thx

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:
```diff
@@ -664,16 +664,16 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
+      new CalendarInterval(15, 9, -3 * CalendarInterval.MICROS_PER_HOUR), CalendarIntervalType),
       StringType),
-      "interval 1 years 3 months -3 days")
+      "interval 1 years 3 months 1 weeks 2 days -3 hours")
```
@MaxGekk @dongjoon-hyun this is another instance we can think of: I feel `9 days` is more readable than `1 week 2 days`.

cloud-fan commented:
LGTM except a few minor comments, thanks for your work and patience @LinhongLiu !

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17955/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -2702,7 +2715,10 @@ object Sequence {
          |    $t = org.apache.spark.sql.catalyst.util.DateTimeUtils.timestampAddInterval(
-         |       $startMicros, $i * $stepMonths, $i * $stepMicros, $zid);
+         |       $startMicros,
+         |       new org.apache.spark.unsafe.types.CalendarInterval(
```
Sorry I missed this place. It has performance penalty if we create an object every time. Let's go back to the previous version. Sorry for the back and forth!

LinhongLiu replied:
np. I reset the changes back.

SparkQA commented:
**[Test build #113034 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113034/testReport)** for PR 26134 at commit [`fb1591e`](https://github.com/apache/spark/commit/fb1591ebcc8574e3edc465d7d23a2ab4f74df2a1).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113034/
Test FAILed.

LinhongLiu committed:
address comments

SparkQA commented:
**[Test build #113061 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113061/testReport)** for PR 26134 at commit [`2dd8e43`](https://github.com/apache/spark/commit/2dd8e432ed2183db052e8a3ffb4465af745c5e79).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17982/
Test PASSed.

SparkQA commented:
**[Test build #113061 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113061/testReport)** for PR 26134 at commit [`2dd8e43`](https://github.com/apache/spark/commit/2dd8e432ed2183db052e8a3ffb4465af745c5e79).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113061/
Test FAILed.

SparkQA commented:
**[Test build #113073 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113073/testReport)** for PR 26134 at commit [`2f90189`](https://github.com/apache/spark/commit/2f901894eb24d7deec31947708b9afdaae2a4866).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17993/
Test PASSed.

SparkQA commented:
**[Test build #113073 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113073/testReport)** for PR 26134 at commit [`2f90189`](https://github.com/apache/spark/commit/2f901894eb24d7deec31947708b9afdaae2a4866).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113073/
Test PASSed.

LinhongLiu commented:
@cloud-fan tests are all pass, could you please check this PR again?

cloud-fan commented:
thanks, merging to master!

LinhongLiu commented:
thanks @cloud-fan @xuanyuanking @MaxGekk  for your reviewing.

zsxwing commented on common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java:
```diff
@@ -31,45 +32,50 @@
   }
 
-  public CalendarInterval(int months, long microseconds) {
+  public CalendarInterval(int months, int days, long microseconds) {
```
@LinhongLiu Could you send out a follow up PR to document why we need `days` and how do we use it in this file? This is definitely worth to document. Otherwise, everyone reading this class may need to git blame and go through the long discussion in this PR.

LinhongLiu replied:
> @LinhongLiu Could you send out a follow up PR to document why we need `days` and how do we use it in this file? This is definitely worth to document. Otherwise, everyone reading this class may need to git blame and go through the long discussion in this PR.

Sure, will do.
